test/e2e/ip_reuse.go

```diff
@@ -0,0 +1,163 @@
+	bmv4helper, err := patch.NewHelper(&baremetalv4Pool[0], managementClusterClient)
+	Expect(err).NotTo(HaveOccurred())
+	baremetalv4Pool[0].Spec.PreAllocations = baremetalv4PoolNameIndexes
+	Expect(bmv4helper.Patch(ctx, &baremetalv4Pool[0])).To(Succeed())
+
+	By("Patch provisioning IPPool with new allocations using Preallocations field")
+	Expect(managementClusterClient.Get(ctx, client.ObjectKey{Namespace: input.Namespace, Name: provisioningPoolName}, &provisioningPool[0])).To(Succeed())
+	provhelper, err := patch.NewHelper(&provisioningPool[0], managementClusterClient)
+	Expect(err).NotTo(HaveOccurred())
+	provisioningPool[0].Spec.PreAllocations = provisioningPoolNameIndexes
+	Expect(provhelper.Patch(ctx, &provisioningPool[0])).To(Succeed())
+
+	// patch MD maxUnavailable and k8s version to start a parallel upgrade
+	By("Get MachineDeployment")
+	machineDeployments := framework.GetMachineDeploymentsByCluster(ctx, framework.GetMachineDeploymentsByClusterInput{
+		Lister:      managementClusterClient,
+		ClusterName: input.ClusterName,
+		Namespace:   input.Namespace,
+	})
+	Expect(len(machineDeployments)).To(Equal(1), "Expected exactly 1 MachineDeployment")
+	md := machineDeployments[0]
+
+	Byf("Update MachineDeployment maxUnavailable to number of replicas and k8s version from %s to %s", fromK8sVersion, kubernetesVersion)
+	patch := []byte(fmt.Sprintf(`{
+		"spec": {
+			"strategy": {
+				"rollingUpdate": {
+					"maxSurge": 0,
+					"maxUnavailable": 3
+				}
+			},
+			"template": {
+				"spec": {
+					"version": "%s"
+				}
+			}
+		}
+	}`, kubernetesVersion))
+	err = managementClusterClient.Patch(ctx, md, client.RawPatch(types.MergePatchType, patch))
+	Expect(err).To(BeNil(), "Failed to patch MachineDeployment")
+
+	Byf("Wait until %d BMH(s) in deprovisioning state", 3)
+	WaitForNumBmhInState(ctx, bmov1alpha1.StateDeprovisioning, WaitForNumInput{
+		Client:    managementClusterClient,
+		Options:   []client.ListOption{client.InNamespace(input.Namespace)},
+		Replicas:  3,
+		Intervals: input.E2EConfig.GetIntervals(input.SpecName, "wait-bmh-deprovisioning"),
+	})
+
+	Byf("Wait until all %d machine(s) become(s) running", 4)
+	WaitForNumMachinesInState(ctx, clusterv1.MachinePhaseRunning, WaitForNumInput{
+		Client:    managementClusterClient,
+		Options:   []client.ListOption{client.InNamespace(input.Namespace)},
+		Replicas:  4,
+		Intervals: input.E2EConfig.GetIntervals(input.SpecName, "wait-machine-running"),
+	})
+
+	By("Get the IPPools in the cluster")
+	baremetalv4Pool, provisioningPool = GetIPPools(ctx, managementClusterClient, input.ClusterName, input.Namespace)
+	Expect(baremetalv4Pool).To(HaveLen(1))
+	Expect(provisioningPool).To(HaveLen(1))
+
+	By("Check if same IP addresses are reused for nodes")
+	bmv4equal := reflect.DeepEqual(baremetalv4Pool[0].Spec.PreAllocations, baremetalv4Pool[0].Status.Allocations)
```

Comment:
how about printing out in both cases: the `spec` and `status` of IPPools while comparing to the console to make it visible and easier for debugging in the future?